tool_path_planner/src/plane_slicer_raster_generator.cpp

```diff
@@ -62,6 +62,65 @@ struct RasterConstructData
   std::vector<double> segment_lengths;
 };
 
+// @brief this function accepts and returns a rasterConstruct where every segment progresses in the same direction and
+// in the right order
+static RasterConstructData alignRasterCD(const RasterConstructData rcd)
+{
+  if (rcd.raster_segments.size() == 1)
+    return (rcd);  // do nothing if only one segment
+  // determine raster direction from 1st segment, and origin as first point in first raster (raster_start)
+  Eigen::Vector3d raster_start, raster_end, raster_direction;
+  rcd.raster_segments[0]->GetPoint(0, raster_start.data());
+  rcd.raster_segments[0]->GetPoint(rcd.raster_segments[0]->GetPoints()->GetNumberOfPoints() - 1, raster_end.data());
+  raster_direction = raster_end - raster_start;
+  raster_direction.normalize();
+
+  // determine location and direction of each successive segement, reverse any mis-directed segments
+  RasterConstructData temp_rcd;
+  std::list<std::tuple<double, size_t> > seg_order;  // once sorted this list will be the order of the segments
+  std::tuple<double, size_t> p(0.0, 0);
+  seg_order.push_back(p);
+  for (size_t i = 1; i < rcd.raster_segments.size(); i++)
+  {
+    // determine i'th segments direction
+    Eigen::Vector3d seg_start, seg_end, seg_dir;
+    rcd.raster_segments[i]->GetPoint(0, seg_start.data());
+    rcd.raster_segments[i]->GetPoint(rcd.raster_segments[i]->GetPoints()->GetNumberOfPoints() - 1, seg_end.data());
```

Comment:
We should probably check number of points before grabbing the 0th and (N-1)th points.  If there are zero points, the -1th point will not exist, and if there is only 1 point, this code may not behave well.  I know we shouldn't be getting many/any 0- or 1-length segments, but it's better to have these checks than cause segfaults later.

Reply:
> If there are zero points, the -1th point will not exist

Of course, if there are 0 points, the 0th point will not exist either ... not sure how that slipped by me and I latched on to '-1th'.